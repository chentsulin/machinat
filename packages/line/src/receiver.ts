import crypto from 'crypto';
import invariant from 'invariant';
import { makeClassProvider } from '@machinat/core/service';
import type { PopEventWrapper } from '@machinat/core/types';
import WebhookReceiver from '@machinat/http/webhook';
import type { WebhookHandler } from '@machinat/http/webhook/types';

import eventFactory from './event/factory';
import { BotP } from './bot';
import { LINE } from './constant';
import { ConfigsI, PlatformMounterI } from './interface';
import type { LineWebhookRequestBody, LineEventContext } from './types';

type LineReceiverOptions = {
  providerId: string;
  channelId: string;
  shouldValidateRequest?: boolean;
  channelSecret?: string;
};

/** @internal */
const handleWebhook = (
  {
    providerId,
    channelId,
    shouldValidateRequest,
    channelSecret,
  }: LineReceiverOptions,
  bot: BotP,
  popEventWrapper: PopEventWrapper<LineEventContext, null>
): WebhookHandler => {
  const popEvent = popEventWrapper(() => Promise.resolve(null));

  return async (metadata) => {
    const {
      request: { method, body, headers },
    } = metadata;
    if (method !== 'POST') {
      return { code: 405 };
    }

    if (!body) {
      return { code: 400 };
    }

    if (shouldValidateRequest && channelSecret !== undefined) {
      const signature = crypto
        .createHmac('SHA256', channelSecret)
        .update(body)
        .digest('base64');

      if (headers['x-line-signature'] !== signature) {
        return { code: 401 };
      }
    }

    let parsedBody: LineWebhookRequestBody;
    try {
      parsedBody = JSON.parse(body);
    } catch (e) {
      return { code: 400 };
    }

    const { events } = parsedBody;
    if (!events) {
      return { code: 400 };
    }

    const issuingEvents: Promise<null>[] = [];

    for (const rawEvent of events) {
      const event = eventFactory(providerId, channelId, rawEvent);

      issuingEvents.push(
        popEvent({
          platform: LINE,
          bot,
          event,
          metadata,
        })
      );
    }

    await Promise.all(issuingEvents);
    return { code: 200 };
  };
};

/**
 * @category Provider
 */
export class LineReceiver extends WebhookReceiver {
  constructor(
    {
      shouldValidateRequest = true,
      channelId,
      providerId,
      channelSecret,
    }: LineReceiverOptions,
    bot: BotP,
    popEventWrapper: PopEventWrapper<LineEventContext, null>
  ) {
    invariant(providerId, 'configs.providerId should not be empty');
    invariant(channelId, 'configs.channelId should not be empty');
    invariant(
      !shouldValidateRequest || channelSecret,
      'should provide configs.channelSecret when shouldValidateRequest set to true'
    );

    super(
      handleWebhook(
        { shouldValidateRequest, providerId, channelId, channelSecret },
        bot,
        popEventWrapper
      )
    );
  }
}

export const ReceiverP = makeClassProvider({
  lifetime: 'singleton',
  deps: [ConfigsI, BotP, PlatformMounterI] as const,
  factory: (configs, bot, { popEventWrapper }) =>
    new LineReceiver(configs, bot, popEventWrapper),
})(LineReceiver);

export type ReceiverP = LineReceiver;
