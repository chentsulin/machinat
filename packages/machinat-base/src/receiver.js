// @flow
import type {
  EventIssuer,
  MachinatChannel,
  MachinatEvent,
  MachinatMetadata,
} from './types';

// BaseReceiver provide event/error issuer binding logic for implementation
// classes to inherit.
class BaseReceiver<
  Channel: MachinatChannel,
  Event: MachinatEvent<any>,
  Metadata: MachinatMetadata<any>,
  Response
> {
  _isBound: boolean;
  _issueEvent: EventIssuer<Channel, Event, Metadata, Response>;
  _issueError: Error => void;

  constructor() {
    this._isBound = false;
  }

  get isBound() {
    return this._isBound;
  }

  bindIssuer(
    eventIssuer: EventIssuer<Channel, Event, Metadata, Response>,
    errorIssuer: (err: Error) => void
  ): boolean {
    if (this._isBound) {
      return false;
    }

    this._isBound = true;
    this._issueEvent = eventIssuer;
    this._issueError = errorIssuer;
    return true;
  }

  unbindIssuer(): boolean {
    if (!this._isBound) {
      return false;
    }

    this._isBound = false;
    return true;
  }
}

export default BaseReceiver;
