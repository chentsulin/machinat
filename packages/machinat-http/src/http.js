// @flow
import url from 'url';
import type { IncomingMessage, ServerResponse } from 'http';
import { HTTPReceivable } from './types';

export const connect = (bot: HTTPReceivable) => (
  req: IncomingMessage,
  res: ServerResponse
) => {
  bot.adaptor.handleRequest(req, res);
};

export const connectByRoutes = (mapping: { [string]: HTTPReceivable }) => {
  const routes = Object.keys(mapping);

  return (req: IncomingMessage, res: ServerResponse) => {
    const parsed = url.parse(req.url);
    const pathname = parsed.pathname || '';

    for (const route of routes) {
      if (pathname.slice(-route.length) === route) {
        mapping[route].adaptor.handleRequest(req, res);
        return;
      }
    }

    res.statusCode = 404; // eslint-disable-line no-param-reassign
    res.end();
  };
};
