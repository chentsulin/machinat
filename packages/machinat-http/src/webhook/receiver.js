// @flow
import type { IncomingMessage, ServerResponse } from 'http';
import readRawBody from 'raw-body';
import type { WebhookHandler } from './types';

const RAW_BODY_OPTION = { encoding: true };

const objectHasOwnProperty = (obj, prop) =>
  Object.prototype.hasOwnProperty.call(obj, prop);

class WebhookReceiver {
  _handleWebhook: WebhookHandler;

  constructor(webhookHandler: WebhookHandler) {
    this._handleWebhook = webhookHandler;
  }

  handleRequestCallback() {
    return this.handleRequest.bind(this);
  }

  async handleRequest(req: IncomingMessage, res: ServerResponse) {
    try {
      let body;
      if (
        req.method === 'POST' ||
        req.method === 'PUT' ||
        req.method === 'DELETE' ||
        req.method === 'PATCH'
      ) {
        body = await readRawBody(req, RAW_BODY_OPTION);
      }

      const metadata = {
        source: 'webhook',
        request: {
          method: req.method,
          url: req.url,
          headers: req.headers,
          body,
        },
      };

      const { code, headers, body: resBody } = await this._handleWebhook(
        metadata
      );

      res.writeHead(code, headers);
      res.end(
        typeof resBody === 'string'
          ? resBody
          : resBody
          ? JSON.stringify(resBody)
          : undefined
      );
    } catch (err) {
      res.statusCode = 500;
      res.end(
        JSON.stringify({
          code: 500,
          message: err.message,
        })
      );
    }
  }
}

export default WebhookReceiver;
