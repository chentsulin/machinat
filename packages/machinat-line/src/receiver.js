// @flow
import crypto from 'crypto';
import invariant from 'invariant';
import { provider } from '@machinat/core/service';
import type { PopEventWrapper } from '@machinat/core/types';
import WebhookReceiver from '@machinat/http/webhook';
import type { WebhookHandler } from '@machinat/http/webhook/types';

import createEvent from './event';
import LineChannel from './channel';
import LineUser from './user';
import LineBot from './bot';
import { LINE } from './constant';
import { LINE_PLATFORM_CONFIGS_I, LINE_PLATFORM_MOUNTER_I } from './interface';
import type {
  LineWebhookRequestBody,
  LineEventContext,
  LinePlatformConfigs,
  LinePlatformMounter,
} from './types';

type LineReceiverOptions = {
  providerId: string,
  botChannelId: string,
  shouldValidateRequest?: boolean,
  channelSecret?: string,
};

const handleWebhook = (
  {
    providerId,
    botChannelId,
    shouldValidateRequest,
    channelSecret,
  }: LineReceiverOptions,
  bot: LineBot,
  popEventWrapper: PopEventWrapper<LineEventContext, null>
): WebhookHandler => {
  const popEvent = popEventWrapper(() => Promise.resolve(null));

  return async (metadata) => {
    const {
      request: { method, body, headers },
    } = metadata;
    if (method !== 'POST') {
      return { code: 405 };
    }

    if (!body) {
      return { code: 400 };
    }

    if (shouldValidateRequest && channelSecret !== undefined) {
      const signature = crypto
        .createHmac('SHA256', channelSecret)
        .update(body)
        .digest('base64');

      if (headers['x-line-signature'] !== signature) {
        return { code: 401 };
      }
    }

    let parsedBody;
    try {
      parsedBody = (JSON.parse(body): LineWebhookRequestBody);
    } catch (e) {
      return { code: 400 };
    }

    const { events } = parsedBody;
    if (!events) {
      return { code: 400 };
    }

    const issuingEvents = [];

    for (const rawEvent of events) {
      const { source } = rawEvent;
      const channel = LineChannel.fromMessagingSource(
        providerId,
        botChannelId,
        source
      );
      const user = new LineUser(providerId, botChannelId, source.userId);
      const event = createEvent(rawEvent);

      issuingEvents.push(
        popEvent({
          platform: LINE,
          bot,
          channel,
          user,
          event,
          metadata,
        })
      );
    }

    await Promise.all(issuingEvents);
    return { code: 200 };
  };
};

class LineReceiver extends WebhookReceiver {
  constructor(
    {
      shouldValidateRequest = true,
      botChannelId,
      providerId,
      channelSecret,
    }: LineReceiverOptions,
    bot: LineBot,
    popEventWrapper: PopEventWrapper<LineEventContext, null>
  ) {
    invariant(
      !shouldValidateRequest || channelSecret,
      'should provide channelSecret if shouldValidateRequest set to true'
    );

    super(
      handleWebhook(
        { shouldValidateRequest, providerId, botChannelId, channelSecret },
        bot,
        popEventWrapper
      )
    );
  }
}

export default provider<LineReceiver>({
  lifetime: 'singleton',
  deps: [LINE_PLATFORM_CONFIGS_I, LineBot, LINE_PLATFORM_MOUNTER_I],
  factory: (
    configs: LinePlatformConfigs,
    bot: LineBot,
    { popEventWrapper }: LinePlatformMounter
  ) => new LineReceiver(configs, bot, popEventWrapper),
})(LineReceiver);
