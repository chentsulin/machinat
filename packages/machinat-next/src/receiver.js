// @flow
import url from 'url';
import type { IncomingMessage, ServerResponse } from 'http';
import { BaseReceiver } from 'machinat-base';
import type { HTTPRequestReceiver } from 'machinat-http-adaptor/types';
import type { NextEvent, NextMetadata, NextParams } from './types';

const NEXT_SERVER_CHANNEL = {
  platform: 'next',
  type: 'server',
  uid: 'next:server',
};

export type NextChannel = typeof NEXT_SERVER_CHANNEL;

class NextReceiver
  extends BaseReceiver<NextChannel, NextEvent, NextMetadata, NextParams>
  implements HTTPRequestReceiver {
  _next: Object;

  constructor(next: Object) {
    super();
    this._next = next;
  }

  handleRequest(req: IncomingMessage, res: ServerResponse) {
    this._handleRequestImpl(req, res);
  }

  callback() {
    return this.handleRequest.bind(this);
  }

  async _handleRequestImpl(req: IncomingMessage, res: ServerResponse) {
    const next = this._next;
    const { pathname, query } = url.parse(req.url, true);

    try {
      if (!this.isBound) {
        res.statusCode = 501; // eslint-disable-line no-param-reassign
        await next.renderError(null, req, res, pathname, query);
        return;
      }

      const response = await this._issueEvent(
        NEXT_SERVER_CHANNEL,
        {
          platform: 'next',
          type: 'request',
          payload: {
            pathname: pathname || '',
            query: query || {},
          },
        },
        {
          source: 'next',
          request: {
            method: req.method,
            url: req.url,
            headers: req.headers,
            encrypted: !!(req: any).connection.encrypted,
          },
        }
      );

      if (!response) {
        res.statusCode = 501; // eslint-disable-line no-param-reassign
        await next.renderError(null, req, res, pathname, query);
        return;
      }

      await next.render(req, res, response.pathname, response.query);
    } catch (err) {
      this._issueError(err);
      await next.renderError(err, req, res, pathname, query);
    }
  }
}

export default NextReceiver;
