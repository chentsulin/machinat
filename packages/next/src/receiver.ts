import { parse as parseUrl, UrlWithParsedQuery } from 'url';
import { STATUS_CODES, IncomingMessage, ServerResponse } from 'http';
import {
  maybeInjectContainer,
  makeClassProvider,
  createEmptyScope,
  ServiceScope,
} from '@machinat/core/service';
import ModuleUtilitiesI from '@machinat/core/base/ModuleUtilities';
import type { PopErrorFn } from '@machinat/core';
import type { RequestHandler } from '@machinat/http';
import { ServerI, ConfigsI } from './interface';
import type { NextServer, NextRequestHandler } from './types';

type NextReceiverOptions = {
  entryPath?: string;
  noPrepare?: boolean;
  handleRequest?: NextRequestHandler;
  initScope?: () => ServiceScope;
  popError?: PopErrorFn | null;
};

/**
 * @category Provider
 */
export class NextReceiver {
  private _next: NextServer;
  private _defaultNextHandler: (
    req: IncomingMessage,
    res: ServerResponse,
    parsedUrl: UrlWithParsedQuery
  ) => void;

  private _pathPrefix: string;
  private _shouldPrepare: boolean;
  private _prepared: boolean;

  private _requestHandler: NextRequestHandler;
  private _initScope: () => ServiceScope;
  private _popError: (err: Error) => void;

  constructor(
    nextApp: NextServer,
    {
      noPrepare,
      entryPath,
      handleRequest,
      initScope,
      popError,
    }: NextReceiverOptions
  ) {
    this._next = nextApp;
    this._defaultNextHandler = nextApp.getRequestHandler();
    this._pathPrefix = entryPath ? entryPath.replace(/\/$/, '') : '';
    this._shouldPrepare = !noPrepare;
    this._prepared = false;

    this._requestHandler = handleRequest || (() => ({ ok: true }));
    this._initScope = initScope || createEmptyScope;
    this._popError =
      popError ||
      (() => {
        // do nothing
      });
  }

  async prepare(): Promise<void> {
    if (this._shouldPrepare) {
      await this._next.prepare();
    }
    this._prepared = true;
  }

  handleRequest(req: IncomingMessage, res: ServerResponse): void {
    this._handleRequestImpl(req, res).catch(this._popError);
  }

  handleRequestCallback(): RequestHandler {
    return this.handleRequest.bind(this);
  }

  private async _handleRequestImpl(req: IncomingMessage, res: ServerResponse) {
    if (!this._prepared) {
      res.statusCode = 503; // eslint-disable-line no-param-reassign
      res.end(STATUS_CODES[503]);
      return;
    }

    const parsedUrl = parseUrl(req.url as string, true);
    const pathPrefix = this._pathPrefix;

    const { pathname } = parsedUrl;
    const trimedPath = !pathname
      ? undefined
      : pathPrefix === ''
      ? pathname
      : pathname.indexOf(pathPrefix) === 0
      ? pathname.slice(pathPrefix.length) || '/'
      : undefined;

    if (!trimedPath) {
      res.statusCode = 404;
      await this._next.renderError(
        null,
        req,
        res,
        pathname as string,
        parsedUrl.query
      );
      return;
    }

    const parsedUrlWithPathPrefixTrimed = {
      ...parsedUrl,
      pathname: trimedPath,
      path: (parsedUrl.path as string).slice(pathPrefix.length) || '/',
    };

    if (trimedPath.slice(1, 6) === '_next') {
      if (this._next.renderOpts.dev && pathPrefix !== '') {
        // HACK: to make react hot loader server recognize the request in
        //       dev environment
        // eslint-disable-next-line no-param-reassign
        req.url = trimedPath;
      }

      this._defaultNextHandler(req, res, parsedUrlWithPathPrefixTrimed);
      return;
    }

    try {
      const request = {
        method: req.method as string,
        url: req.url as string,
        headers: req.headers,
      };

      const response = await maybeInjectContainer(
        this._initScope(),
        this._requestHandler
      )(request);

      if (response.ok) {
        const { page, query, headers } = response;
        if (headers) {
          for (const [key, value] of Object.entries(headers)) {
            res.setHeader(key, value as string);
          }
        }

        if (page || query) {
          await this._next.render(
            req,
            res,
            page || trimedPath,
            query || parsedUrl.query,
            parsedUrlWithPathPrefixTrimed
          );
        } else {
          this._defaultNextHandler(req, res, parsedUrlWithPathPrefixTrimed);
        }
      } else {
        const { code, reason, headers } = response;
        res.writeHead(code, headers);
        await this._next.renderError(
          new Error(reason),
          req,
          res,
          trimedPath,
          parsedUrl.query
        );
      }
    } catch (err) {
      this._popError(err);

      res.statusCode = 500;
      await this._next.renderError(err, req, res, trimedPath, parsedUrl.query);
    }
  }
}

export const ReceiverP = makeClassProvider({
  lifetime: 'singleton',
  deps: [
    ServerI,
    ConfigsI,
    { require: ModuleUtilitiesI, optional: true },
  ] as const,
  factory: (nextApp, { entryPath, noPrepare, handleRequest }, moduleUtils) =>
    new NextReceiver(nextApp, {
      entryPath,
      noPrepare,
      handleRequest,
      initScope: moduleUtils?.initScope,
      popError: moduleUtils?.popError,
    }),
})(NextReceiver);

export type ReceiverP = NextReceiver;
