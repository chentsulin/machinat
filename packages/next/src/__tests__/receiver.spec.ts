import {
  createEmptyScope,
  makeContainer,
  ServiceScope,
} from '@machinat/core/service';
import moxy from '@moxyjs/moxy';
import type createNextApp from 'next';
import { IncomingMessage, ServerResponse } from 'http';
import { NextReceiver } from '../receiver';

const nextDefaultHandler = moxy();

const nextApp = moxy<ReturnType<typeof createNextApp>>({
  getRequestHandler: () => nextDefaultHandler,
  prepare: async () => {},
  render: async () => {},
  renderError: async () => {},
  setAssetPrefix() {},
  renderOpts: { assetPrefix: '' },
} as never);

const nextTick = () => new Promise((resolve) => process.nextTick(resolve));

const requestHandler = moxy(() => ({ ok: true as const }));
const initScope = moxy(createEmptyScope);
const popError = moxy();

const req = moxy<IncomingMessage>({
  method: 'GET',
  url: 'http://machinat.com/hello?foo=bar',
  headers: { 'x-y-z': 'abc' },
} as never);
let res;

beforeEach(() => {
  nextDefaultHandler.mock.clear();
  nextApp.mock.reset();
  req.mock.reset();
  res = moxy(new ServerResponse(req));

  requestHandler.mock.reset();
  initScope.mock.reset();
  popError.mock.reset();
});

it('respond 503 if request received before prepared', async () => {
  const receiver = new NextReceiver(nextApp, {
    noPrepare: false,
    handleRequest: requestHandler,
  });

  expect(receiver.handleRequest(req, res)).toBe(undefined);
  await nextTick();

  expect(res.statusCode).toBe(503);
  expect(res.end.mock).toHaveBeenCalledTimes(1);

  expect(nextApp.renderError.mock).not.toHaveBeenCalled();
  expect(requestHandler.mock).not.toHaveBeenCalled();
});

test('#prepare() call next.prepare()', async () => {
  const receiver = new NextReceiver(nextApp, { noPrepare: false });
  expect(nextApp.prepare.mock).not.toHaveBeenCalled();

  await receiver.prepare();

  expect(nextApp.prepare.mock).toHaveBeenCalledTimes(1);
});

test('#prepare() not call next.prepare() if shouldPrepare is false', async () => {
  const receiver = new NextReceiver(nextApp, { noPrepare: true });
  await receiver.prepare();

  expect(nextApp.prepare.mock).not.toHaveBeenCalled();
});

test('default handleRequest behavior', async () => {
  const receiver = new NextReceiver(nextApp, { noPrepare: true, popError });
  await receiver.prepare();

  expect(receiver.handleRequest(req, res)).toBe(undefined);
  await nextTick();

  expect(res.statusCode).toBe(200);

  expect(nextApp.renderError.mock).not.toHaveBeenCalled();
  expect(nextApp.render.mock).not.toHaveBeenCalled();

  expect(nextDefaultHandler.mock).toHaveBeenCalledTimes(1);
  expect(nextDefaultHandler.mock).toHaveBeenCalledWith(
    req,
    res,
    expect.any(Object)
  );
  expect(nextDefaultHandler.mock.calls[0].args[2]).toMatchInlineSnapshot(`
    Object {
      "auth": null,
      "hash": null,
      "host": "machinat.com",
      "hostname": "machinat.com",
      "href": "http://machinat.com/hello?foo=bar",
      "path": "/hello?foo=bar",
      "pathname": "/hello",
      "port": null,
      "protocol": "http:",
      "query": Object {
        "foo": "bar",
      },
      "search": "?foo=bar",
      "slashes": true,
    }
  `);
});

it('trim url path with entryPath', async () => {
  const receiver = new NextReceiver(nextApp, {
    noPrepare: true,
    entryPath: '/hello',
    popError,
    handleRequest: requestHandler,
  });
  await receiver.prepare();

  req.mock.getter('url').fakeReturnValue('http://machinat.com/hello/world');
  receiver.handleRequest(req, res);
  await nextTick();

  expect(nextApp.renderError.mock).not.toHaveBeenCalled();
  expect(nextApp.render.mock).not.toHaveBeenCalled();

  expect(nextDefaultHandler.mock).toHaveBeenCalledTimes(1);
  expect(nextDefaultHandler.mock).toHaveBeenCalledWith(
    req,
    res,
    expect.any(Object)
  );

  const parsedUrlPassed = nextDefaultHandler.mock.calls[0].args[2];
  expect(parsedUrlPassed.path).toBe('/world');
  expect(parsedUrlPassed.pathname).toBe('/world');
});

it('respond 404 if entryPath not match', async () => {
  const receiver = new NextReceiver(nextApp, {
    noPrepare: true,
    entryPath: '/hello',
    popError,
    handleRequest: requestHandler,
  });
  await receiver.prepare();

  req.mock.getter('url').fakeReturnValue('http://machinat.com/foo/world');

  receiver.handleRequest(req, res);
  await nextTick();

  expect(requestHandler.mock).not.toHaveBeenCalled();
  expect(nextApp.render.mock).not.toHaveBeenCalled();
  expect(nextDefaultHandler.mock).not.toHaveBeenCalled();

  expect(nextApp.render.mock).not.toHaveBeenCalled();
  expect(nextApp.renderError.mock).toHaveBeenCalledTimes(1);
  expect(nextApp.renderError.mock).toHaveBeenCalledWith(
    null,
    req,
    res,
    '/foo/world',
    {}
  );

  expect(res.mock.setter('statusCode')).toHaveBeenCalledTimes(1);
  expect(res.mock.setter('statusCode')).toHaveBeenCalledWith(404);
});

test('customized headers returned by handleRequest', async () => {
  requestHandler.mock.fake(async () => ({
    ok: true,
    headers: { 'x-foo': 'foo', 'x-bar': 'bar' },
  }));

  const receiver = new NextReceiver(nextApp, {
    noPrepare: true,
    popError,
    handleRequest: requestHandler,
  });
  await receiver.prepare();

  expect(receiver.handleRequest(req, res)).toBe(undefined);
  await nextTick();

  expect(requestHandler.mock).toHaveBeenCalledTimes(1);
  expect(requestHandler.mock.calls[0].args[0]).toMatchSnapshot();

  expect(res.statusCode).toBe(200);
  expect(res.getHeaders()).toEqual({ 'x-foo': 'foo', 'x-bar': 'bar' });
});

it('call next.render() with page/query returned by handleRequest', async () => {
  const receiver = new NextReceiver(nextApp, {
    noPrepare: true,
    popError,
    handleRequest: requestHandler,
  });
  await receiver.prepare();

  requestHandler.mock.fake(async () => ({
    ok: true,
    page: '/hello/world',
    query: { foo: 'bar' },
    headers: { 'x-y-z': 'a_b_c' },
  }));

  expect(receiver.handleRequest(req, res)).toBe(undefined);
  await nextTick();

  expect(requestHandler.mock).toHaveReturnedTimes(1);
  expect(requestHandler.mock.calls[0].args[0]).toMatchSnapshot();

  expect(nextDefaultHandler.mock).not.toHaveBeenCalled();
  expect(nextApp.renderError.mock).not.toHaveBeenCalled();

  expect(nextApp.render.mock).toHaveBeenCalledTimes(1);
  expect(nextApp.render.mock).toHaveBeenCalledWith(
    req,
    res,
    '/hello/world',
    { foo: 'bar' },
    expect.any(Object)
  );

  expect(nextApp.render.mock.calls[0].args[4]).toMatchInlineSnapshot(`
    Object {
      "auth": null,
      "hash": null,
      "host": "machinat.com",
      "hostname": "machinat.com",
      "href": "http://machinat.com/hello?foo=bar",
      "path": "/hello?foo=bar",
      "pathname": "/hello",
      "port": null,
      "protocol": "http:",
      "query": Object {
        "foo": "bar",
      },
      "search": "?foo=bar",
      "slashes": true,
    }
  `);

  const expectedRequestObj = {
    method: 'GET',
    url: 'http://machinat.com/hello?foo=bar',
    headers: { 'x-y-z': 'abc' },
  };

  expect(requestHandler.mock).toHaveBeenCalledTimes(1);
  expect(requestHandler.mock).toHaveBeenCalledWith(expectedRequestObj);

  expect(res.statusCode).toBe(200);
  expect(res.getHeaders()).toEqual({ 'x-y-z': 'a_b_c' });
});

test('use service container for handleRequest', async () => {
  const handleRequestFn = moxy(() => ({ ok: true as const, page: '/foo' }));
  const useRequestHandler = moxy(() => handleRequestFn);

  const receiver = new NextReceiver(nextApp, {
    noPrepare: true,
    popError,
    initScope,
    handleRequest: makeContainer({ deps: [ServiceScope] })(useRequestHandler),
  });
  await receiver.prepare();

  expect(receiver.handleRequest(req, res)).toBe(undefined);
  await nextTick();

  expect(initScope.mock).toHaveBeenCalledTimes(1);
  expect(useRequestHandler.mock).toHaveReturnedTimes(1);
  expect(useRequestHandler.mock).toHaveBeenCalledWith(
    initScope.mock.calls[0].result
  );

  expect(nextDefaultHandler.mock).not.toHaveBeenCalled();
  expect(nextApp.renderError.mock).not.toHaveBeenCalled();

  expect(nextApp.render.mock).toHaveBeenCalledTimes(1);
  expect(nextApp.render.mock).toHaveBeenCalledWith(
    req,
    res,
    '/foo',
    { foo: 'bar' },
    expect.any(Object)
  );

  expect(handleRequestFn.mock).toHaveBeenCalledTimes(1);
  expect(handleRequestFn.mock).toHaveBeenCalledWith({
    method: 'GET',
    url: 'http://machinat.com/hello?foo=bar',
    headers: { 'x-y-z': 'abc' },
  });

  expect(res.statusCode).toBe(200);
});

it('call next.renderError() handleRequest return not ok', async () => {
  const receiver = new NextReceiver(nextApp, {
    noPrepare: true,
    popError,
    handleRequest: requestHandler,
  });
  await receiver.prepare();

  requestHandler.mock.fake(async () => ({
    ok: false,
    code: 418,
    reason: "I'm a teapot",
  }));

  expect(receiver.handleRequest(req, res)).toBe(undefined);
  await nextTick();

  expect(requestHandler.mock).toHaveBeenCalledTimes(1);
  expect(requestHandler.mock.calls[0].args[0]).toMatchSnapshot();

  expect(nextApp.render.mock).not.toHaveBeenCalled();
  expect(nextDefaultHandler.mock).not.toHaveBeenCalled();

  expect(nextApp.renderError.mock).toHaveBeenCalledTimes(1);
  expect(nextApp.renderError.mock).toHaveBeenCalledWith(
    new Error("I'm a teapot"),
    req,
    res,
    '/hello',
    { foo: 'bar' }
  );

  expect(res.mock.setter('statusCode')).toHaveBeenCalledWith(418);
  expect(popError.mock).not.toHaveBeenCalled();
});

it('call next.renderError() with customized headers return by handleRequest', async () => {
  const receiver = new NextReceiver(nextApp, {
    noPrepare: true,
    popError,
    handleRequest: requestHandler,
  });
  await receiver.prepare();

  requestHandler.mock.fake(async () => ({
    ok: false,
    code: 418,
    reason: "I'm a teapot",
    headers: { 'x-x-x': 't-e-a-p-o-t' },
  }));

  expect(receiver.handleRequest(req, res)).toBe(undefined);
  await nextTick();

  expect(nextApp.renderError.mock).toHaveBeenCalledWith(
    new Error("I'm a teapot"),
    req,
    res,
    '/hello',
    { foo: 'bar' }
  );

  expect(res.writeHead.mock).toHaveBeenCalledWith(418, {
    'x-x-x': 't-e-a-p-o-t',
  });
});

it('pass "_next" api calls to next directly', async () => {
  const receiver = new NextReceiver(nextApp, {
    noPrepare: true,
    entryPath: '/hello',
    popError,
    handleRequest: requestHandler,
  });
  await receiver.prepare();

  req.mock.getter('url').fakeReturnValue('http://machinat.com/hello/_next/xxx');

  receiver.handleRequest(req, res);
  await nextTick();

  expect(requestHandler.mock).not.toHaveBeenCalled();
  expect(nextApp.render.mock).not.toHaveBeenCalled();
  expect(nextApp.renderError.mock).not.toHaveBeenCalled();

  expect(nextDefaultHandler.mock).toHaveBeenCalledTimes(1);
  expect(nextDefaultHandler.mock).toHaveBeenCalledWith(
    req,
    res,
    expect.any(Object)
  );
  const parsedUrl = nextDefaultHandler.mock.calls[0].args[2];
  expect(parsedUrl.path).toBe('/_next/xxx');
  expect(parsedUrl.pathname).toBe('/_next/xxx');
});

test('entryPath does not affect page params from middlewares', async () => {
  const receiver = new NextReceiver(nextApp, {
    noPrepare: true,
    entryPath: '/hello',
    popError,
    handleRequest: requestHandler,
  });
  await receiver.prepare();

  requestHandler.mock.fake(async () => ({
    ok: true,
    page: '/hello/world',
    query: { foo: 'bar' },
  }));

  expect(receiver.handleRequest(req, res)).toBe(undefined);
  await nextTick();

  expect(nextApp.render.mock).toHaveBeenCalledTimes(1);
  expect(nextApp.render.mock).toHaveBeenCalledWith(
    req,
    res,
    '/hello/world',
    { foo: 'bar' },
    expect.any(Object)
  );
});

it('call next.renderError() if middlewares reject', async () => {
  const receiver = new NextReceiver(nextApp, {
    noPrepare: true,
    popError,
    handleRequest: requestHandler,
  });
  await receiver.prepare();

  requestHandler.mock.fake(async () => {
    throw new Error("I'm a teapot");
  });

  expect(receiver.handleRequest(req, res)).toBe(undefined);
  await nextTick();

  expect(requestHandler.mock).toHaveBeenCalledTimes(1);

  expect(nextApp.render.mock).not.toHaveBeenCalled();
  expect(nextDefaultHandler.mock).not.toHaveBeenCalled();

  expect(nextApp.renderError.mock).toHaveBeenCalledTimes(1);
  expect(nextApp.renderError.mock).toHaveBeenCalledWith(
    new Error("I'm a teapot"),
    req,
    res,
    '/hello',
    { foo: 'bar' }
  );

  expect(res.mock.setter('statusCode')).toHaveBeenCalledWith(500);

  expect(popError.mock).toHaveBeenCalledTimes(1);
  expect(popError.mock).toHaveBeenCalledWith(new Error("I'm a teapot"));
});
