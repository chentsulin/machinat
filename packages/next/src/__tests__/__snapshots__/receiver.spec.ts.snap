// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`call next.render() with page/query returned by handleRequest 1`] = `
Object {
  "headers": Object {
    "x-y-z": "abc",
  },
  "method": "GET",
  "url": "http://machinat.com/hello?foo=bar",
}
`;

exports[`call next.renderError() handleRequest return not ok 1`] = `
Object {
  "headers": Object {
    "x-y-z": "abc",
  },
  "method": "GET",
  "url": "http://machinat.com/hello?foo=bar",
}
`;

exports[`customized headers returned by handleRequest 1`] = `
Object {
  "headers": Object {
    "x-y-z": "abc",
  },
  "method": "GET",
  "url": "http://machinat.com/hello?foo=bar",
}
`;
